Add recursive-analysis sub-command

stash86 opened the pull request:
<!-- Thank you for sending your pull request. But first, have you included
unit tests, and is your code PEP8 conformant? [More details](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md)
-->
## Summary
Add recursive-analysis sub-command to test indicators of potential recursive issue (value changes when startup candle change)

Still to-do:
1. [x] Tests
2. [x] Documentations
3. [x] Maybe combine it to lookahead-analysis

Solve the issue: #___

stash86 committed:
add recursive analysis

stash86 committed:
add mising const

stash86 committed:
add missing arg

stash86 committed:
fix ruff

stash86 committed:
pre-commit fixes

stash86 committed:
ruff fix

stash86 committed:
fix mypy

stash86 committed:
fix mypy

stash86 committed:
fix flake8

stash86 committed:
Merge branch 'freqtrade:develop' into bt-metrics

xmatthias commented on freqtrade/optimize/recursive_analysis.py:
```diff
@@ -0,0 +1,224 @@
+                                            str(self.dt_to_timestamp(varholder.to_dt)))
+        prepare_data_config['exchange']['pair_whitelist'] = pairs_to_load
+
+        backtesting = Backtesting(prepare_data_config, self.exchange)
```
What exactly (from backtesting) do you need? 
My initial assumption was that it's only analyzing indicators (?) which might not even need a backtesting class (?)

stash86 replied:
good question. I just took hippo's code and modify them, so not sure if I can do things without calling backtesting. On quick glance, at least I need backtesting to get the data to calculate indicators

xmatthias replied:
leave this for now - i'll look at it in detail

in theory, you shouldn't - plotting also doesn't need it (and i don't think you need funding rates or detail timeframe - as you don't really run a backtest afterwards)

xmatthias reviewed (commented):
from a high level view, the whole `RecursiveAnalysis` class seems to share a great deal of code with lookahead-bias (not everything obviously - but there's functions that are  90% identical, if not more).


i think some refactoring will greatly reduce complexity / code duplication (which undoubly will cause deviation over time) - and with that increase maintainability greatly.

`RecursiveAnalysisSubFunctions` might be a similar candidate, too - though it's doesn't seem to share that much code - at least it's not immediately obvious

stash86 commented:
Options that I have in mind
1. Combine my code to hippo's code, and add flag to switch between lookahead and recursive
2. Create new class as main class that hold those similar functions

xmatthias commented:
well clearly option 2
option 1 would/will result in too much "if recursive" - they're somewhat identical - but not THAT similar  :laughing:

stash86 committed:
Merge branch 'freqtrade:develop' into bt-metrics

stash86 committed:
create BaseAnalysis class

stash86 committed:
add missing var

stash86 committed:
debug

stash86 committed:
add missing args to config

stash86 committed:
update lookahead analysis

stash86 committed:
fix mutable Backtest

stash86 committed:
remove prepare data from baseanalysis

stash86 committed:
precommit fix

stash86 committed:
flake8 fix

stash86 committed:
more fixes

stash86 committed:
fix flake8

stash86 committed:
Merge pull request #20 from stash86/bt-metrics2

create BaseAnalysis class

stash86 committed:
Merge branch 'freqtrade:develop' into bt-metrics2

stash86 committed:
fix wording

stash86 committed:
Merge branch 'freqtrade:develop' into bt-metrics

stash86 committed:
Merge branch 'freqtrade:develop' into bt-metrics2

stash86 committed:
Merge remote-tracking branch 'origin/bt-metrics2' into bt-metrics2

stash86 committed:
initial doc

stash86 committed:
remove 1 column

stash86 committed:
fix docs

stash86 committed:
fix example in the docs, increasing startup to 400 on ema100

stash86 committed:
change startup cande in sample strat

stash86 committed:
initial test

stash86 committed:
1 test

stash86 committed:
first test done

stash86 committed:
4 more tests

stash86 committed:
all tests done

stash86 committed:
pre-commit fix

stash86 committed:
Merge branch 'bt-metrics2' of https://github.com/stash86/freqtrade into bt-metrics

stash86 committed:
fix test after adding new test strategy

stash86 committed:
fix another test

stash86 committed:
fix test

stash86 commented:
tests and docs are done. I have created BaseAnalysis to put similar functions there (not many).

stash86 committed:
add more explanation in the docs about startup candle and its relation to API limit

froggleston committed:
Update recursive-analysis.md

I went through the docs which are great, but they needed some tidying up and changes to language to meet the style of the rest of the docs.

froggleston reviewed (changes requested):
This looks really good!

My only suggestion would be to include an example output table in the docs to help users understand what the different output values mean.

stash86 committed:
add example of the result of the analysis

froggleston committed:
Update recursive-analysis.md

Thanks for adding the example table! It's good to have that example available. I've edited this section of the docs as before.

xmatthias commented on freqtrade/templates/sample_strategy.py:
```diff
@@ -77,7 +77,7 @@ class SampleStrategy(IStrategy):
 
     # Number of candles the strategy requires before producing valid signals
-    startup_candle_count: int = 30
+    startup_candle_count: int = 170
```
I assume this is the outcome of running the analysis against this strategy?

stash86 replied:
Kinda. It's supposed to be 200 actually. Gonna update it

xmatthias commented on freqtrade/commands/arguments.py:
```diff
@@ -467,3 +470,14 @@ def _build_subcommands(self) -> None:
+        # Add recursive_analysis subcommand
+        recursive_analayis_cmd = subparsers.add_parser(
+            'recursive-analysis',
+            help="Check for potential look ahead bias.",
```
I think we might need a better help string.
This is identical to lookahead_analysis - but the command does not do exactly that.

xmatthias commented on docs/recursive-analysis.md:
```diff
@@ -0,0 +1,91 @@
+# Recursive analysis
```
We'll need to decide what to do with this.
The way it's currently configured, it'll NOT show up in the left-side nav - and will therefore only be reachable either through direct links - or through the link in strategy-customization.

i'm ... tempted to combine this with "lookahead analysis" - as the commands are strongly related, and i think we already have a "too long" nav ...

stash86 replied:
Personally, I felt like Backtesting, Hyperopt, Lookahead and Recursive analysis should all be combined into 1 main menu.

xmatthias replied:
you mean something like this

```
Backtesting
    Backtesting
    hyperopt
    lookahead analysis
    recursive analysis
```

not against this per se - though i don't actually like the behavior of expanding boxes, as it'd require 2 clicks to actually go to backtesting (for example).

stash86 replied:
another place I have in mind is `Advanced Topics`

xmatthias replied:
but you'd keep it as 2 separate pages? or do something similar than what we do with plugins ... ?

stash86 replied:
keep as separate pages.

xmatthias committed:
Fix docs rendering (lists need a newline before the first item)

xmatthias commented on docs/recursive-analysis.md:
```diff
@@ -0,0 +1,92 @@
+head over to the [Data Downloading](data-download.md) section of the documentation.
+
+This command is built upon backtesting since it internally chains backtests to prepare different lengths of data and calculates indicators based on the downloaded data.
+This does not run the strategy itself, but rather uses the indicators it contains. After multiple backtests are done to calculate the indicators of different startup candle values (`startup_candle_count`), the values of last rows across all backtests are compared to see how much variance they show compared to the base backtest.
```
i think we need to work on wording here (thoughout this doc).
while it's using the backtest class currently - it's not actually executing a backtest - but runs the populate_* methods (actually only `populate_indicators()`).

a "backtest" would assume a full run (`backtest.start()`) - which ain't what's actually happening ... 
We might also need a warning someplace that indicators calculated in `populate_entry_trend()` are ignored for this analysis.

stash86 committed:
fix wrong startup candle

stash86 committed:
fix args description

stash86 committed:
move lookahead and recursive docs under Advanced Topics

stash86 committed:
change wordings on the doc from backtest to "calculate indicators" since we never do full backtest process

stash86 committed:
fix pre-commit

xmatthias committed:
Improve sorting in docs

(these are more important than advanced strategy/docs)

xmatthias committed:
Fix typo in --strategy_list

xmatthias committed:
Rename test to avoid naming collision

xmatthias commented on freqtrade/optimize/recursive_analysis_helpers.py:
```diff
@@ -0,0 +1,106 @@
+            # setting a timerange is enforced here
+            raise OperationalException(
+                "Please set a timerange. "
+                "A timerange of 20 candles are enough for recursive analysis."
```
interesting, this seems to contradict to what we're saying in the docs ... i think we need to adjust either this, or the docs (i think the docs are ... more correct in this regard).

> It is recommended to set a long timerange (at least 5000 candles) so that the initial indicators' calculation that is going to be used as a benchmark has very small or no recursive issues itself. For example, for a 5m timeframe, a timerange of 5000 candles would be equal to 18 days.

stash86 replied:
Ah yes. On it rn

xmatthias committed:
Add "bias2" test with full lookahead bias

xmatthias commented on freqtrade/optimize/recursive_analysis.py:
```diff
@@ -0,0 +1,182 @@
+            # print(compare_df)
+            for col_name, values in compare_df.items():
+                # print(col_name)
+                if 'other' == col_name:
+                    continue
```
what's the reason to exclude `'other'` from this test ? 

`dataframe['other'] = ta.RSI(dataframe, timeperiod=50).shift(-1)` is ok - but `dataframe['rsi'] = ta.RSI(dataframe, timeperiod=50).shift(-1)` is not? :D

stash86 replied:
we skip 'other' because `.compare` will create 2 columns  'self' and 'others', and we only need to go through it just once

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.compare.html

xmatthias replied:
ah ok perfect - thanks for the explanation

xmatthias committed:
Further update test a bit

xmatthias committed:
Slightly updated docs

stash86 committed:
fix error message

stash86 committed:
Merge remote-tracking branch 'origin/bt-metrics' into bt-metrics

xmatthias reviewed (approved):
We might want to do some deduplication of the 2 helper methods - but i'm intentionally leaving this for a future enhancement.
This command is too valueable to not merge now.

LGTM :+1:

froggleston commented:
LET'S GOOOOOOOOOOOO! Great addition!
